File: 프로그래머스/3/43164. 여행경로/여행경로.java
// 완전탐색 -> 모든 가능한 경로 구하기
// 정렬(알파벳 순서가 앞서는 경로)
// 가장 첫번째 요소 리턴
import java.util.*;

class Solution {
    static boolean[] visited; // 방문 체크
    static List<String>result = new ArrayList<>();
    
    public String[] solution(String[][] tickets) {
        String[] answer = {};
        // 변수세팅
        visited = new boolean[tickets.length];
        
        // 완탐 및 정렬
        dfs(tickets, "ICN", "ICN", 0);
        
        Collections.sort(result);
        
        String[] ans = result.get(0).split(" ");
        
        return ans;
    }
    public void dfs(String[][] tickets, String start, String route, int idx){
        // 종료 조건
        if(idx==tickets.length){
            result.add(route);
            return;
        }
        // 백트래킹
        for(int i=0; i<tickets.length; i++){
            if(tickets[i][0].equals(start) && !visited[i]){
                visited[i]=true;
                dfs(tickets, tickets[i][1],route+" "+tickets[i][1], idx+1 );
                visited[i]=false;
            }
        }
        return;
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.py
# 방문하는 공항 경로 배열에 담아 return
# ICN에서 출발
def solution(tickets):
    answer = []
    
    graph = {} # 딕셔너리로 저장
    
    for start, end in tickets:
        if start in graph:
            graph[start].append(end)
        else:
            graph[start]=[end]
            
    # 알파벳 순서로 정렬
    for airport in graph:
        graph[airport].sort()
    # 경로 찾기
    route =[]
    
    def dfs(airport):
        # 현재 공항이 graph의 key로 존재, 다음 공항이 있다면
        while airport in graph and graph[airport]:
            next = graph[airport].pop(0) # 다음 경로 찾기
            dfs(next)
        route.append(airport)
        
    dfs("ICN")
    
    return route[::-1]